src/librustc_codegen_ssa/mir/constant.rs

```diff
@@ -40,29 +40,37 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
         &mut self,
         constant: &mir::Constant<'tcx>,
     ) -> Result<ConstValue<'tcx>, ErrorHandled> {
-        match constant.literal.val {
-            ty::ConstKind::Unevaluated(def_id, substs, promoted) => {
-                let substs = self.monomorphize(&substs);
-                self.cx
-                    .tcx()
-                    .const_eval_resolve(ty::ParamEnv::reveal_all(), def_id, substs, promoted, None)
-                    .map_err(|err| {
-                        if promoted.is_none() {
-                            self.cx
-                                .tcx()
-                                .sess
-                                .span_err(constant.span, "erroneous constant encountered");
-                        }
-                        err
-                    })
-            }
-            ty::ConstKind::Value(value) => Ok(value),
-            _ => {
-                let const_ = self.monomorphize(&constant.literal);
-                if let ty::ConstKind::Value(value) = const_.val {
-                    Ok(value)
-                } else {
-                    span_bug!(constant.span, "encountered bad ConstKind in codegen: {:?}", const_);
+        // repeatedly monomorphize generic parameters until we either
+        // reach a value or an error.
```

Comment:
This is in the wrong direction. You should *never* monomorphize more than once.
If the result isn't monomorphic, *something went very wrong*.

Reply:
@davidtwco and I recently spent some time tracking down multiple monomorphizations, as they broke down their "polymorphization" work (right now nothing happens if you apply the same substitutions more than once in codegen, because the substitutions themselves are monomorphic, but that doesn't make it *correct*, and it certainly breaks down with polymorphization).